File: cs/usage-maven-plugin/src/main/java/com/usalko/maven/plugin/usage/asm/UsageMethodVisitor.java
package com.usalko.maven.plugin.usage.asm;

/*
 * Copyright 2001-2005 The Apache Software Foundation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import org.objectweb.asm.AnnotationVisitor;
import org.objectweb.asm.Attribute;
import org.objectweb.asm.Handle;
import org.objectweb.asm.Label;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.TypePath;

class UsageMethodVisitor extends MethodVisitor {

    private final VisitorsContext visitorsContext;

    UsageMethodVisitor(VisitorsContext visitorsContext) {
        super(Opcodes.ASM7);
        this.visitorsContext = visitorsContext;
    }

    @Override
    public void visitMethodInsn(int opcode, String owner, String name, String desc, boolean itf) {
        super.visitMethodInsn(opcode, owner, name, desc, itf);
        visitorsContext.acceptClass(owner);
    }

    @Override
    public AnnotationVisitor visitAnnotationDefault() {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public AnnotationVisitor visitTypeAnnotation(int typeRef, TypePath typePath, String descriptor,
            boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public void visitAttribute(Attribute attribute) {
        super.visitAttribute(attribute);
        visitorsContext.acceptClass(attribute.type);
    }

    @Override
    public void visitTypeInsn(int opcode, String type) {
        super.visitTypeInsn(opcode, type);
        visitorsContext.acceptClass(type);
    }

    @Override
    public void visitFieldInsn(int opcode, String owner, String name, String descriptor) {
        super.visitFieldInsn(opcode, owner, name, descriptor);
        visitorsContext.acceptClass(owner);
    }

    @Override
    public void visitInvokeDynamicInsn(String name, String descriptor, Handle bootstrapMethodHandle,
            Object... bootstrapMethodArguments) {
        super.visitInvokeDynamicInsn(name, descriptor, bootstrapMethodHandle,
                bootstrapMethodArguments);
        visitorsContext.acceptClass(descriptor);
    }

    @Override
    public void visitMultiANewArrayInsn(String descriptor, int numDimensions) {
        super.visitMultiANewArrayInsn(descriptor, numDimensions);
        visitorsContext.acceptClass(descriptor);
    }

    @Override
    public AnnotationVisitor visitInsnAnnotation(int typeRef, TypePath typePath, String descriptor,
            boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public void visitTryCatchBlock(Label start, Label end, Label handler, String type) {
        super.visitTryCatchBlock(start, end, handler, type);
        visitorsContext.acceptClass(type);
    }

    @Override
    public AnnotationVisitor visitTryCatchAnnotation(int typeRef, TypePath typePath,
            String descriptor, boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public void visitLocalVariable(String name, String descriptor, String signature, Label start,
            Label end, int index) {
        super.visitLocalVariable(name, descriptor, signature, start, end, index);
        visitorsContext.acceptClass(descriptor);
    }

    @Override
    public AnnotationVisitor visitLocalVariableAnnotation(int typeRef, TypePath typePath,
            Label[] start, Label[] end, int[] index, String descriptor, boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

}

File: cs/gradle-usage-plugin/src/main/kotlin/com/usalko/gradle/plugin/asm/UsageMethodVisitor.kt
package com.usalko.gradle.plugin.asm

/*
 * Copyright 2001-2005 The Apache Software Foundation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import org.objectweb.asm.*

internal class UsageMethodVisitor(private val visitorsContext: VisitorsContext) : MethodVisitor(Opcodes.ASM7) {

    override fun visitMethodInsn(opcode: Int, owner: String, name: String, desc: String, itf: Boolean) {
        super.visitMethodInsn(opcode, owner, name, desc, itf)
        visitorsContext.acceptClass(owner)
    }

    override fun visitAnnotationDefault(): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitAnnotation(descriptor: String, visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitTypeAnnotation(typeRef: Int, typePath: TypePath?, descriptor: String,
                                     visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitAttribute(attribute: Attribute) {
        super.visitAttribute(attribute)
        visitorsContext.acceptClass(attribute.type)
    }

    override fun visitTypeInsn(opcode: Int, type: String?) {
        super.visitTypeInsn(opcode, type)
        visitorsContext.acceptClass(type)
    }

    override fun visitFieldInsn(opcode: Int, owner: String, name: String, descriptor: String) {
        super.visitFieldInsn(opcode, owner, name, descriptor)
        visitorsContext.acceptClass(owner)
    }

    override fun visitInvokeDynamicInsn(name: String, descriptor: String, bootstrapMethodHandle: Handle,
                                        vararg bootstrapMethodArguments: Any) {
        super.visitInvokeDynamicInsn(name, descriptor, bootstrapMethodHandle,
                *bootstrapMethodArguments)
        visitorsContext.acceptClass(descriptor)
    }

    override fun visitMultiANewArrayInsn(descriptor: String, numDimensions: Int) {
        super.visitMultiANewArrayInsn(descriptor, numDimensions)
        visitorsContext.acceptClass(descriptor)
    }

    override fun visitInsnAnnotation(typeRef: Int, typePath: TypePath?, descriptor: String,
                                     visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitTryCatchBlock(start: Label, end: Label, handler: Label, type: String?) {
        super.visitTryCatchBlock(start, end, handler, type)
        visitorsContext.acceptClass(type)
    }

    override fun visitTryCatchAnnotation(typeRef: Int, typePath: TypePath?,
                                         descriptor: String, visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitLocalVariable(name: String, descriptor: String, signature: String?, start: Label,
                                    end: Label, index: Int) {
        super.visitLocalVariable(name, descriptor, signature, start, end, index)
        visitorsContext.acceptClass(descriptor)
    }

    override fun visitLocalVariableAnnotation(typeRef: Int, typePath: TypePath?,
                                              start: Array<Label>, end: Array<Label>,
                                              index: IntArray, descriptor: String,
                                              visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

}
